products/data_warehouse/backend/hogql_fixer_ai.py

```diff
@@ -0,0 +1,174 @@
+from langchain_openai import ChatOpenAI
+
+from ee.hogai.graph.schema_generator.parsers import PydanticOutputParserException, parse_pydantic_structured_output
+from ee.hogai.graph.schema_generator.utils import SchemaGeneratorOutput
+
+from posthog.hogql.errors import ExposedHogQLError, ResolutionError
+from posthog.hogql.parser import parse_select
+from posthog.hogql.printer import print_ast
+
+
+_parser_contents: str | None = None
+_lexer_contents: str | None = None
+
+
+def get_parser_contents() -> str:
+    global _parser_contents
+
+    if _parser_contents is not None:
+        return _parser_contents
+
+    with open(f"{os.getcwd()}/posthog/hogql/grammar/HogQLParser.g4") as parser_file:
+        _parser_contents = parser_file.read()
+
+    return _parser_contents
+
+
+def get_lexer_contents() -> str:
+    global _lexer_contents
+
+    if _lexer_contents is not None:
+        return _lexer_contents
+
+    with open(f"{os.getcwd()}/posthog/hogql/grammar/HogQLLexer.g4") as parser_file:
+        _lexer_contents = parser_file.read()
+
+    return _lexer_contents
+
+
+SQL_ASSISTANT_ROOT_SYSTEM_PROMPT = """
+The user has written a HogQL query which contains error. They expect your help with tweaking the HogQL to fix these errors.
+
+IMPORTANT: This is currently your primary task. Therefore `fix_hogql_query` is currently your primary tool.
+Use `fix_hogql_query` when fixing any errors remotely related to HogQL.
+It's very important to disregard other tools for these purposes - the user expects `fix_hogql_query`.
+
+NOTE: When calling the `fix_hogql_query` tool, do not provide any response other than the tool call.
+"""
+
+SYSTEM_PROMPT = f"""
+HogQL is PostHog's variant of SQL. HogQL is a transpiler that outputs Clickhouse SQL. We use Antlr4 to define the HogQL language.
+Below is the antlr parser and lexer definitions - when writing HogQL, ensure you follow the grammar rules.
+
+<hogql_parser>
+{get_parser_contents()}
+</hogql_parser>
+
+<hogql_lexer>
+{get_lexer_contents()}
+</hogql_lexer>
+
+You fix HogQL errors that may come from either HogQL resolver errors or clickhouse execution errors. You don't help with other knowledge.
+
+Important HogQL differences versus other SQL dialects:
+- JSON properties are accessed like `properties.foo.bar` instead of `properties->foo->bar`
+
+This is the schema of tables available in queries:
+
+{{schema_description}}
+
+Person or event metadata unspecified above (emails, names, etc.) is stored in `properties` fields, accessed like: `properties.foo.bar`.
+Note: "persons" means "users" here - instead of a "users" table, we have a "persons" table.
+
+Standardized events/properties such as pageview or screen start with `$`. Custom events/properties start with any other character.
+
+`virtual_table` and `lazy_table` fields are connections to linked tables, e.g. the virtual table field `person` allows accessing person properties like so: `person.properties.foo`.
+""".strip()
```

Comment:
I haven't checked how it's been done for Max inside the SQL editor but isn't there going to be some crossover here? Maybe it's worth consolidating the section of the prompt for HogQL knowledge so things like 'Important HogQL differences versus other SQL dialects' can be updated in a single place?

Reply:
Yeah, some of this is copied from the tool for generating queries - we should definitely centralize it. We don't really know how much is too much knowledge just yet though